List files & directories for collection & page APIs

mertkahyaoglu opened the pull request:
Current version of `collection` & `page` APIs list all of the files at the root level. This makes it hard to view/edit files in nested folders in the front end. Other than the user experience, we already have issues like #150, #163, #186 showing that it is not possible to edit files in folders right now. 

This PR proposes listing files & directories all together at a requested level. For example, if we send a request like the following;

**GET** `collections/posts/entries/test`

It returns all directories and files under `test` directory;

![screen shot 2017-01-19 at 11 31 48 pm](https://cloud.githubusercontent.com/assets/7414026/22124599/302921f2-de99-11e6-9449-f93ef10dc15c.png)

And the front-end listing view at `posts/test` becomes like this;

![screen shot 2017-01-20 at 12 06 22 am](https://cloud.githubusercontent.com/assets/7414026/22125332/ed0ec1c6-de9b-11e6-9bb6-4083f8670b03.png)

I believe this API model will make it easy to fix aforementioned issues and also provide a better experience for users.

Current status of this PR;

- [x] List files & directories for collections
- [x] List files & directories for pages
- [x] API tests
- [x] Apply changes to the front end
- [x] Front-end tests
- [x] Documentation
- [ ] ~List files & directories for drafts~ (leaving this for another PR)

/cc @benbalter @parkr 

Looking forward to your thoughts on this to move forward confidently.

Closes #163 #150

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,62 @@
+
+    TYPE = 'directory'
+
+    def initialize(path, args={})
```
I *believe* you're trying to use named arguments here... this can be simplified to:

```ruby
def initialize(path, base: nil, content_type: nil, splat: nil)
  @base = Pathname.new base
  @content_type = content_type
  @splat = Pathname.new splat
  @path = Pathname.new path
end
```

mertkahyaoglu replied:
That's good to know, but I need to pass these arguments [here](https://github.com/jekyll/jekyll-admin/blob/455329fa8d811d0b7a2b11dc032b2a1c7643d75e/lib/jekyll-admin/directory.rb#L58) recursively. Is there a shortcut for that in Ruby?

benbalter replied:
`base: base, content_type: content_type, splat: splat`

mertkahyaoglu replied:
These will be undefined since we are outside of `initialize` method.

mertkahyaoglu replied:
I think I should make them `attr_reader`

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,62 @@
+    end
+
+    def name
+      @path.basename
```
1. Since you have `attr_reader :path` above, this can be just `path.basename`

2. To DRY things up a bit, you can add the following in place of the `name` and `modified_time` methods:

```ruby
  extend Forwardable
  def_delegator :@path, :basename, :name
  def_delegator :@path, :mtime, :modified_time
```

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,62 @@
+      else
+        "/collections/#{@content_type}/entries/#{@splat}/#{name}"
+      end
+    end
```
Is this method used?

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,62 @@
+
+    def url
+      resource_path
+    end
```
`alias_method :url, :resource_path`

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,62 @@
+    end
+
+    def http_url
+    end
```
What's your intention here?

mertkahyaoglu replied:
Since directories do not render to html, I wanted to make it `null` in the resulted JSON.

benbalter commented on lib/jekyll-admin/server/collection.rb:
```diff
@@ -59,14 +65,44 @@ def document
+        args = {
+          base: site.source,
+          content_type: params['collection_id'],
+          splat: params["splat"].first
```
Instead of splat, is this really path?

mertkahyaoglu replied:
I need this to fill the gap in the `api_url`. For example, if the requested page in `test/other` directory relative to `pages`, I need to be able to generate the `api_url` like this;

/_api/pages/entries/**test/other**/test.md

mertkahyaoglu commented:
@benbalter I applied the changes you requested, added more tests. 
In addition to them, I changed the `sinatra` routes for pages & collections in order to support directories. Could you review them as well?

For the front end;

- Applied API changes to pages & collections, now you can edit/view the files inside subdirectories
- Added `Title` & `Path` labels to edit view
- Reimplemented `Breadcrumbs` component so that its only task now is to show links. I extracted the input for `path` into another component called `InputPath`.
- Added more tests

The new look;

![screen shot 2017-02-02 at 5 12 35 pm](https://cloud.githubusercontent.com/assets/7414026/22552998/caf4d904-e963-11e6-9da7-22568539ea9e.png)


@benbalter I've encountered an issue while renaming files. Sometimes, after writing to the disc, the resulted page/document might be `nil` which causes 404. As far as I see, this happens randomly. Sometimes it renames and returns the renamed file, sometimes it returns nil. So I couldn't figure out what the problem actually is. This is valid for both pages & collections.

@ashmaroli , @DirtyF  Guys, your feedbacks about the user interface are appreciated very much. 😃

benbalter commented on lib/jekyll-admin/apiable.rb:
```diff
@@ -38,20 +37,26 @@ def to_api(include_content: false, include_documents: false)
+
+      if is_a?(Jekyll::Document)
+        output["dir"] = File.join("/", relative_path_from_source.dirname.to_s)
+        output["name"] = relative_path_from_source.basename
```
For non-Documents, this should already be calculated as `name`. For documents, this should be already calculated as `basename_without_ext`.

mertkahyaoglu replied:
Didn't know that. Thanks!

benbalter commented on lib/jekyll-admin/apiable.rb:
```diff
@@ -38,20 +37,26 @@ def to_api(include_content: false, include_documents: false)
+      end
+
+      if is_a?(Jekyll::Document)
+        output["dir"] = File.join("/", relative_path_from_source.dirname.to_s)
```
I'd add a helper method for this. For non-Documents, this should already be calculated as `@dir`. For Documents, it should be already calculated as `relative_path`.

mertkahyaoglu replied:
Yeap, but `relative_path` includes `filename` while `dir` does not. However, I thought I'd need this in the front end but I didn't. I'm removing this field.

benbalter commented on lib/jekyll-admin/apiable.rb:
```diff
@@ -38,20 +37,26 @@ def to_api(include_content: false, include_documents: false)
 
+    # returns the relative path from site's source
+    def relative_path_from_source
+      Pathname.new(file_path).relative_path_from(Pathname.new(site.source))
```
I *believe* this should be the value of `@dir` or `relative_path`.

mertkahyaoglu replied:
Just noticed this isn't used anywhere 😄

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,58 @@
+
+    TYPE = "directory".freeze
+
+    def initialize(path, base: nil, content_type: nil, splat: nil)
```
Can you add a comment explaining what each of these arguments represent?

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,58 @@
+        :path          => relative_path,
+        :type          => TYPE,
+      }
+    end
```
Just making a note to be sure to compare this hash to other data types to ensure the field names are consistent.

mertkahyaoglu replied:
I based these field names on static files which also have `name`, `modified_time` and `path` fields.

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,58 @@
+    include JekyllAdmin::URLable
+    include JekyllAdmin::APIable
+
+    TYPE = "directory".freeze
```
It's a strange ruby concept, but strings that aren't modified should be stored as `:symbols` which are treated differently for garbage collection purposes.

mertkahyaoglu replied:
Good to know that 👍  So I should do this?;

```ruby
TYPE = "directory".to_sym
```

mertkahyaoglu replied:
or directly;

```ruby
{
        :name          => name,
        :modified_time => modified_time,
        :path          => relative_path,
        :type          => "directory",
      }
```

since it is used in only one place.

benbalter replied:
`TYPE = :directory`

benbalter commented on lib/jekyll-admin/directory.rb:
```diff
@@ -0,0 +1,58 @@
+    end
+
+    def resource_path
+      if content_type == "pages"
```
If we're passing this ourselves, it should be a `:symbol`.

mertkahyaoglu replied:
`content_type` passed from `page.rb` as a constant "pages" and from `collection.rb` as a route parameter `collection_id`. Based on the type, `api_url` and `http_url` are generated via `resource_path`. I don't know how this should be `:symbol`

benbalter replied:
If it's coming from the URL, the this is correct as is. You never want to call `to_sym` on user input, as it's a DDoS vector.

benbalter commented on spec/jekyll-admin/server/collection_spec.rb:
```diff
@@ -32,56 +32,74 @@ def app
     end
 
-    it "includes documents" do
+    it "doesn't includes documents" do
```
Can you explain the changes to the API that are happening here and why? I'm having trouble following. If we're changing the API do the docs need to be updated?

mertkahyaoglu replied:
I wanted to get your approval before updating them, but sure I can do it right now.

Basically, we are not exposing documents on `/:collection_name` endpoint anymore. This endpoint has only the information about the collection itself as it had in the first release. This could be beneficial in the long term that we might want to change the collection's attributes such as `label` and get only the necessary fields while doing it.

Related to this PR, we now don't have only documents but also have directories that are listed in index endpoints. Therefore putting them all under a field like `documents` or `entries` along with the collection info makes no sense. Instead we put them under their own endpoints. For example;

GET `/posts`
```javascript
{
"label": "posts",
"files": [],
"directory": "/Users/mertkahyaoglu/dev/jekyll-admin/spec/fixtures/site/_posts",
"relative_directory": "_posts",
"permalink": "/:categories/:year/:month/:day/:title:output_ext",
"http_url": null,
"api_url": "http://localhost:4000/_api/collections/posts"
}
```

GET `/posts/entries`, GET `/posts/entries/test`, etc.

```javascript
[
{
"name": "test",
"modified_time": "2017-02-02 17:51:11 +0200",
"path": "_posts/test",
"type": "directory",
...
},
{
"path": "_posts/2016-03-01-test.md",
"id": "/2016/03/01/test",
"relative_path": "_posts/2016-03-01-test.md",
"url": "/2016/03/01/test.html",
"collection": "posts",
...
}
]
```

mertkahyaoglu commented:
> Sometimes, after writing to the disc, the resulted page/document might be nil which causes 404. As far as I see, this happens randomly. 

@benbalter Any idea why this is happening. I'm producing this in older versions too. Could it be related with `site.process` being late to serve new pages/documents?

benbalter commented on docs/api.md:
```diff
@@ -95,13 +121,15 @@ Delete the requested document from disk.
-Return an array of page objects. The response does not include the page body.
+Returns an array of pages and directories for the requested path. If `path`
+is not provided, entries at the root level are returned. The response does not include
+pages' body.
```
Nice. Does this include directories too (like the collections endpoint)?

mertkahyaoglu replied:
Absolutely 👍  Also fixes #163 since it includes directories like `_pages`.

benbalter commented on lib/jekyll-admin/server/collection.rb:
```diff
@@ -7,32 +7,34 @@ class Server < Sinatra::Base
-        json document.to_api(:include_content => true)
+        write_file(write_path, document_body)
+        updated_document = collection.docs.find { |d| d.path == write_path }
+        render_404 if updated_document.nil?
```
I *think* you should be able to use `ensure_document` here.

mertkahyaoglu replied:
I know this doesn't look good but `ensure_document` uses `document` which relies on `document_path` for comparison. We need `write_path` here to compare with.

benbalter commented on lib/jekyll-admin/server/collection.rb:
```diff
@@ -47,26 +49,70 @@ def collection
       end
 
+      def request_path
+        sanitized_path File.join(site.source, request_payload["path"])
```
This should just be `sanitized_path(request_paylod["path"])`. See https://github.com/jekyll/jekyll-admin/blob/140a1f9f598341700796f4044657ea6808f7ea10/lib/jekyll-admin/server.rb#L53-L55.

mertkahyaoglu replied:
Oh, ok 👍  
The only problem is that this adds extra `/` at the end. Is there a convenient way to delete it?

benbalter replied:
Option 1: Add the slash `path << '/' unless path.end_with?('/')`

Option 2: Remove the slash `path.sub(%r{/$}, '')`

mertkahyaoglu replied:
Didn't need that. Translated others to have `/` at the end using `sanitized_path` 👍

mertkahyaoglu replied:
Thanks again 👍  it became very simple now https://github.com/jekyll/jekyll-admin/pull/281/commits/260ddaea8cdeedac1e908578606839150c9615a2

benbalter commented on lib/jekyll-admin/server/collection.rb:
```diff
@@ -47,26 +49,70 @@ def collection
+          relative_to_collection = File.join(params["splat"].first, filename)
+          File.join(collection.relative_directory, relative_to_collection)
+        end
+      end
```
How is this being used? We have to be careful here with user input... I could set a filename of `../../../etc/passwd`.

mertkahyaoglu replied:
Let's say `request_payload["path"]` for `/:collection_id/*?/?:path.:ext` endpoint is `_posts/some_folder/2017-01-01-test.md`. We need to compare this path with the one generated from `params`fields to check if there exists a rename. Any easy solution to achieve this?

benbalter commented on lib/jekyll-admin/server/page.rb:
```diff
@@ -32,21 +34,79 @@ class Server < Sinatra::Base
       private
 
+      def request_path
+        sanitized_path File.join(site.source, request_payload["path"])
```
`site.source` is unnecessary here, as it's already handled in `sanitized_path`

benbalter commented on lib/jekyll-admin/server/page.rb:
```diff
@@ -32,21 +34,79 @@ class Server < Sinatra::Base
 
+      def directory_pages
+        pages.find_all do |p|
+          File.dirname(File.join(site.source, p.path)) == directory_path
```
I'd use `sanitized_path` rather than `File.join` here to prevent directory traversal.

benbalter commented on lib/jekyll-admin/server/page.rb:
```diff
@@ -32,21 +34,79 @@ class Server < Sinatra::Base
+      end
+
+      def directory_path
+        sanitized_path File.join(site.source, params["splat"].first)
```
Don't need `site.source` here.

benbalter commented:
> Any idea why this is happening?

Sounds like a race condition. `write_file` calls `site.process`, which shouldn't return until the site is processed... all steps being blocking. Any chance incremental regeneration is on? Should we try disabling it by default?

mertkahyaoglu commented:
> Any chance incremental regeneration is on? Should we try disabling it by default?

I tried disabling it from `_config.yml` but no work. Still random errors. Is there any other way to do it?

mertkahyaoglu commented:
@benbalter Looks like `site.process` is not blocking. I put `sleep(4)` after it and everything worked as expected.

mertkahyaoglu commented:
@benbalter You're right. I'm opening an issue about that. Thank you for reviewing this PR 👍
